ca/ca_test.go

```diff
@@ -207,261 +194,527 @@ func setup(t *testing.T) *testCtx {
+	// Limit ourselves to only having one ECDSA issuer, just in case they have
+	// different notAfter dates.
+	cargs.boulderIssuers = cargs.boulderIssuers[:3]
 
-	if len(cert.DNSNames) == 1 {
-		if cert.DNSNames[0] != "not-example.com" {
-			t.Errorf("Improper list of domain names %v", cert.DNSNames)
-		}
-		t.Errorf("Improper list of domain names %v", cert.DNSNames)
+	// Jump to a time just moments before the test issuer expire.
+	future := cargs.boulderIssuers[2].Cert.Certificate.NotAfter.Add(-1 * time.Hour)
+	cargs.clk.Set(future)
+
+	ca, err := cargs.make()
+	if err != nil {
+		t.Fatalf("making test ca: %s", err)
 	}
 
-	if len(cert.Subject.Country) > 0 {
-		t.Errorf("Subject contained unauthorized values: %v", cert.Subject)
+	_, err = ca.IssueCertificate(t.Context(), &capb.IssueCertificateRequest{
+		RegistrationID: 1, OrderID: 1,
+		Csr: ECDSACSR, CertProfileName: "legacy",
+	})
+	if err == nil {
+		t.Fatalf("IssueCertificate(notAfter > issuer.notAfter) succeeded, but want error")
+	}
+	if !errors.Is(err, berrors.InternalServer) {
+		t.Fatalf("IssueCertificate(notAfter > issuer.notAfter) = %T, but want %T", err, berrors.InternalServer)
 	}
 }
 
-// Test failure mode when no issuers are present.
-func TestNoIssuers(t *testing.T) {
+func TestIssueCertificate_InvalidProfile(t *testing.T) {
 	t.Parallel()
-	testCtx := setup(t)
-	sa := &mockSA{}
-	_, err := NewCertificateAuthorityImpl(
-		sa,
-		mockSCTService{},
-		testCtx.pa,
-		nil, // No issuers
-		testCtx.certProfiles,
-		testCtx.serialPrefix,
-		testCtx.maxNames,
-		testCtx.keyPolicy,
-		testCtx.logger,
-		testCtx.metrics,
-		testCtx.fc)
-	test.AssertError(t, err, "No issuers found during CA construction.")
-	test.AssertEquals(t, err.Error(), "must have at least one issuer")
+
+	ca, err := newCAArgs(t).make()
+	if err != nil {
+		t.Fatalf("making test ca: %s", err)
+	}
+
+	for _, tc := range []struct {
+		name    string
+		profile string
+		wantErr string
+	}{
+		{
+			name:    "no profile",
+			profile: "",
+			wantErr: "Incomplete issue certificate request",
+		},
+		{
+			name:    "unrecognized profile",
+			profile: "doesnotexist",
+			wantErr: "incapable of using a profile named",
+		},
+		{
+			name:    "invalid profile name",
+			profile: "🤓",
+			wantErr: "incapable of using a profile named",
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			_, err := ca.IssueCertificate(t.Context(), &capb.IssueCertificateRequest{
+				RegistrationID: 1, OrderID: 1,
+				Csr: ECDSACSR, CertProfileName: tc.profile,
+			})
+			if err == nil {
+				t.Fatalf("IssueCertificate(profile=%q) succeeded, but want error", tc.profile)
+			}
+			if !strings.Contains(err.Error(), tc.wantErr) {
+				t.Fatalf("IssueCertificate(profile=%q) = %q, but want %q", tc.profile, err, tc.wantErr)
+			}
+		})
+	}
 }
 
-// Test issuing when multiple issuers are present.
-func TestMultipleIssuers(t *testing.T) {
+func TestIssueCertificate_ProfileSelection(t *testing.T) {
```

Comment:
I thought: "Wow, this is a much better name than TestIssueCertificateForPrecertificateWithSpecificCertificateProfile". But actually... that test case still exists, and it still tests the inner functions. Is that intentional?

Reply:
Yep, it's intentional. I wanted to leave the old test cases in place as long as the underlying functions they were testing are in place. There's a wacky world out there somewhere in which between this change and the subsequent change landing, someone tries to add another call to issueCertificateForPrecertificate somewhere bad. Leaving the old tests in place makes it safer (or, at least makes _me feel_ like it is safer) to leave the helper functions themselves in place.

The old tests are now redundant and are removed in the follow-up PR which removes the helper functions.